### PR TITLE
Pause and Resume consumer

### LINKED ISSUE #867: Implement Consumers Pause
### Proposed change

Implement the new API for pausing consumers.

The Server PR with related change has been merged here: https://github.com/nats-io/nats-server/pull/5066
Schemas PR, which is probably the best source of information how to implement the feature:
https://github.com/nats-io/jsm.go/pull/522/files

#### ADR Issue
https://github.com/nats-io/nats-architecture-and-design/issues/266

## Details
**Request and response schemas:**
```go
// io.nats.jetstream.api.v1.consumer_pause_request
type JSApiConsumerPauseRequest struct {
	PauseUntil time.Time `json:"pause_until,omitempty"`
}

// io.nats.jetstream.api.v1.consumer_pause_response
type JSApiConsumerPauseResponse struct {
	JSApiResponse
	Paused         bool          `json:"paused"`
	PauseUntil     time.Time     `json:"pause_until"`
	PauseRemaining time.Duration `json:"pause_remaining,omitempty"`
}
```
Consumer info includes 2 new fields:

```go
Paused bool json:"paused,omitempty"
PauseRemaining time.Duration json:"pause_remaining,omitempty"
```
The paused state and time time would need to be persisted to the raft layer such that server restarts would not unpause paused consumers. This is done using the consumer configuration that has a new value:
```go
PauseUntil time.Time json:"pause_until,omitempty"
```
If in doubt, refer to the Server/Schema PR

Feature be available in the server version 2.11

### Hints
Probably the easiest way to implement it, is to look how calls to methods for creating / deleting streams and consumers are done.

### Use case

It is difficult to schedule maintenance on central resources on a large distributed system where 100s or 1000s of clients are accessing data in a stream.

This change allows to pause a Consumer such that it appears healthy but just doesnt deliver any messages.

During the pause maintenance can happen and resources accessed by clients will not be under constant pressure, later the stream can be unpaused and work will continue.

This happen without impacting running clients - other than they would see pending messages in stream info but not get any deliveries.

Applies to push and pull consumers.

### PR DESCRIPTION
**Pull request by scottf**
Resolves https://github.com/nats-io/nats.net/issues/867

Adds support for pausing and resuming consumers.

- The `IJetStreamManagement` now has methods to `jsm.PauseConsumer(stream, consumer, pauseUntil)`, as well as `jsm.ResumeConsumer(stream, consumer)`.
- An example for the above is also added to `src/examples`.
- `ConsumerConfiguration`/builder now has support for setting `pauseUntil` as well. Allowing to set it during consumer creation.

### CODE REVIEW

File: src/NATS.Client/Internals/JsonUtils.cs [hunk 1]
```diff
@@ -144,7 +144,8 @@ public static DateTime AsDate(JSONNode node)
         {
             try
             {
-                return DateTime.Parse(node.Value).ToUniversalTime();
+                DateTime dt = DateTime.Parse(node.Value).ToUniversalTime();
+                return dt.Year == 1 ? DateTime.MinValue : dt; // b/c server sends a value that doesn't quite parse to DateTime.MinValue
```
Comment (mtmk) [inline src/NATS.Client/Internals/JsonUtils.cs]: I don't get this one :-(

**Review by mtmk — COMMENTED** [review]
looks good. just a quick question below.

File: src/NATS.Client/Internals/JsonUtils.cs (hunk 1 above)
Comment (scottf) [reply src/NATS.Client/Internals/JsonUtils.cs]: DateTime.MinValue.ToUniversalTime() resolves to `1/1/0001 5:00:00 AM`

But for whatever reason the value the server sends `0001-01-01T00:00:00Z` when resolved with 
```
DateTime.Parse(node.Value).ToUniversalTime()
```
ends up with `1/2/0001 12:00:00 AM`

So I decdide that anything with year 1 must have meant min value. I can't imagine real server date data will have a date of year 1.

Comment (scottf) [reply src/NATS.Client/Internals/JsonUtils.cs]: I think I'm going to do it a different way and revert this. I can check if paused is false when reading the json, and if it is, then I don't even need to parse the time. This is really an issue with JSONNode

**Review by mtmk — APPROVED** [review]
LGTM

**Pull request merged**